[macOS] Videos on Bing.com cannot successfully enter fullscreen

jernoble opened the pull request:
#### fa5e0c0d9325fdd1efad12e2c41021ec0864cfd1
<pre>
[macOS] Videos on Bing.com cannot successfully enter fullscreen
<a href="https://bugs.webkit.org/show_bug.cgi?id=282668">https://bugs.webkit.org/show_bug.cgi?id=282668</a>
<a href="https://rdar.apple.com/126570526">rdar://126570526</a>

Reviewed by Jean-Yves Avenard.

On Bing.com, the page includes a list of iframe elements whose contents
point to youtube.com embeds. As the user scrolls through the page, script
running on Bing.com notes that the iframe elements have been scrolled out
of the viewport, and tears down their contents. During the fullscreen
transition, the iframe contents are in fullscreen mode and are visible in
the top layer, but the div containing those contents has been scrolled
offscreen. Additionally, the code checking for whether the iframes are
visible runs when the window fires the &quot;scroll&quot; and &quot;resize&quot; events, which
are fired during the fullscreen transition.

Two changes are needed to fix this behavior:

1. Calling WebPageProxy::setPageScale()
has the side effect of scrolling the page back to the origin of (0, 0). This
call is replaced with scalePageRelativeToScrollPosition(), which does not
change the scroll position.

2. Defer firing &quot;scroll&quot; and &quot;resize&quot; events until after &quot;fullscreenchange&quot;
fires. This ensures that the page doesn&apos;t get an intermediate state during
the fullscreen transition in which to check the layout of its contents.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setNeedsDOMWindowResizeEvent):
(WebCore::Document::setNeedsVisualViewportResize):
(WebCore::Document::runResizeSteps):
(WebCore::Document::flushDeferredResizeEvents):
(WebCore::Document::runScrollSteps):
(WebCore::Document::flushDeferredScrollEvents):
(WebCore::Document::updateResizeObservations):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::setAnimatingFullscreen):
(WebCore::FullscreenManager::clear):
(WebCore::FullscreenManager::addPendingScheduledResize): Deleted.
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::deferResizeEvents):
(WebCore::Page::flushDeferredResizeEvents):
(WebCore::Page::deferScrollEvents):
(WebCore::Page::flushDeferredScrollEvents):
* Source/WebCore/page/Page.h:
(WebCore::Page::shouldDeferResizeEvents const):
(WebCore::Page::shouldDeferScrollEvents const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::deferResizeEvents):
(WebKit::WebPageProxy::flushDeferredResizeEvents):
(WebKit::WebPageProxy::deferScrollEvents):
(WebKit::WebPageProxy::flushDeferredScrollEvents):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
(-[WKFullScreenWindowController finishedEnterFullScreenAnimation:]):
(-[WKFullScreenWindowController exitFullScreen]):
(-[WKFullScreenWindowController finishedExitFullScreenAnimationAndExitImmediately:]):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::deferResizeEvents):
(WebKit::WebPage::flushDeferredResizeEvents):
(WebKit::WebPage::deferScrollEvents):
(WebKit::WebPage::flushDeferredScrollEvents):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FullscreenScrollAndResize.mm: Added.
(TestWebKitAPI::TEST(Fullscreen, ScrollAndResize)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:

Canonical link: <a href="https://commits.webkit.org/286501@main">https://commits.webkit.org/286501@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a9c1d103abdff2cdcbef5d84f2cdbdff146e9ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76049 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55078 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28948 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80549 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27315 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78165 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64220 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3374 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59636 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17784 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79116 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49515 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65308 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39993 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46914 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22793 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25642 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68030 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23130 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82010 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3418 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2190 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67854 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3572 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65276 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67164 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16768 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11120 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9236 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3368 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6174 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3389 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/5078 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/4148 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5126,33 +5126,22 @@ void Document::updateViewportUnitsOnResize()
 // https://drafts.csswg.org/cssom-view/#run-the-resize-steps
 void Document::runResizeSteps()
 {
+    if (auto page = this->page(); page && page->shouldDeferResizeEvents())
```
```suggestion
    if (RefPtr page = this->page(); page && page->shouldDeferResizeEvents())
```

jernoble replied:
Per the Safer C++ rules, you don't have to hold a RefPtr on an object if the function being called is "trivial", defined as "the function being called is inlined". Since these accessors are just inlined methods returning a bool ivar, a ref isn't necessary.

jyavenard replied:
I'm more concerned about consistency in this file, than strictly making the static analyzer happy.
There's not a single call to `Document::page()` in Document.cpp that don't use the semantics `RefPtr page = page()` 
Document.cpp has been over several iterations to replace all `auto` with `RefPtr` seems to be a step backward to re-introduce the use of auto. 

If not using RefPtr, the other style use here is more something like `if (page() && page()->method()`

jyavenard replied:
That would become then

```suggestion
    if (page() && page()->shouldDeferResizeEvents())
```

jernoble replied:
I'm not going to adopt this style in the patch. It's needlessly wasteful to call `page()` twice just to achieve "consistency". And the Safer Cpp document explicitly allows for this style of call.

jyavenard commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5203,6 +5197,9 @@ static bool serviceScrollAnimationForScrollableArea(const ScrollableArea* scroll
 // https://drafts.csswg.org/cssom-view/#run-the-scroll-steps
 void Document::runScrollSteps()
 {
+    if (auto page = this->page(); page && page->shouldDeferScrollEvents())
```
```suggestion
    if (RefPtr page = this->page(); page && page->shouldDeferScrollEvents())
```

jernoble replied:
Ditto.

jyavenard replied:
ditto too :)

smfr commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9625,6 +9627,8 @@ void Document::updateResizeObservations(Page& page)
     }
     if (!hasResizeObservers() && !m_resizeObserverForContainIntrinsicSize && !m_contentVisibilityDocumentState)
         return;
+    if (page.shouldDeferResizeEvents())
```
"resize events" and "resize observations" are two separate things (the latter is ResizeObserver). This steps does not relate to resize events.

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3639,6 +3639,26 @@ void WebPage::setLastKnownMousePosition(WebCore::FrameIdentifier frameID, IntPoi
     frame->coreLocalFrame()->eventHandler().setLastKnownMousePosition(eventPoint, globalPoint);
 }
 
+void WebPage::deferResizeEvents()
```
Is this a `startDeferring`?

jyavenard reviewed (approved):
r=me with style comments addressed.

jernoble committed:
[macOS] Videos on Bing.com cannot successfully enter fullscreen
https://bugs.webkit.org/show_bug.cgi?id=282668
rdar://126570526

Reviewed by Jean-Yves Avenard.

On Bing.com, the page includes a list of iframe elements whose contents
point to youtube.com embeds. As the user scrolls through the page, script
running on Bing.com notes that the iframe elements have been scrolled out
of the viewport, and tears down their contents. During the fullscreen
transition, the iframe contents are in fullscreen mode and are visible in
the top layer, but the div containing those contents has been scrolled
offscreen. Additionally, the code checking for whether the iframes are
visible runs when the window fires the "scroll" and "resize" events, which
are fired during the fullscreen transition.

Two changes are needed to fix this behavior:

1. Calling WebPageProxy::setPageScale()
has the side effect of scrolling the page back to the origin of (0, 0). This
call is replaced with scalePageRelativeToScrollPosition(), which does not
change the scroll position.

2. Defer firing "scroll" and "resize" events until after "fullscreenchange"
fires. This ensures that the page doesn't get an intermediate state during
the fullscreen transition in which to check the layout of its contents.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setNeedsDOMWindowResizeEvent):
(WebCore::Document::setNeedsVisualViewportResize):
(WebCore::Document::runResizeSteps):
(WebCore::Document::flushDeferredResizeEvents):
(WebCore::Document::runScrollSteps):
(WebCore::Document::flushDeferredScrollEvents):
(WebCore::Document::updateResizeObservations):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::setAnimatingFullscreen):
(WebCore::FullscreenManager::clear):
(WebCore::FullscreenManager::addPendingScheduledResize): Deleted.
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::deferResizeEvents):
(WebCore::Page::flushDeferredResizeEvents):
(WebCore::Page::deferScrollEvents):
(WebCore::Page::flushDeferredScrollEvents):
* Source/WebCore/page/Page.h:
(WebCore::Page::shouldDeferResizeEvents const):
(WebCore::Page::shouldDeferScrollEvents const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::deferResizeEvents):
(WebKit::WebPageProxy::flushDeferredResizeEvents):
(WebKit::WebPageProxy::deferScrollEvents):
(WebKit::WebPageProxy::flushDeferredScrollEvents):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
(-[WKFullScreenWindowController finishedEnterFullScreenAnimation:]):
(-[WKFullScreenWindowController exitFullScreen]):
(-[WKFullScreenWindowController finishedExitFullScreenAnimationAndExitImmediately:]):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::deferResizeEvents):
(WebKit::WebPage::flushDeferredResizeEvents):
(WebKit::WebPage::deferScrollEvents):
(WebKit::WebPage::flushDeferredScrollEvents):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FullscreenScrollAndResize.mm: Added.
(TestWebKitAPI::TEST(Fullscreen, ScrollAndResize)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:

Canonical link: https://commits.webkit.org/286501@main
